File: src/main/java/ch/obermuhlner/astro/javafx/ProgressDialog.java
package ch.obermuhlner.astro.javafx;

import javafx.concurrent.Task;
import javafx.scene.Scene;
import javafx.scene.control.Label;
import javafx.scene.control.ProgressIndicator;
import javafx.scene.layout.BorderPane;
import javafx.stage.Modality;
import javafx.stage.Stage;
import javafx.stage.StageStyle;

public class ProgressDialog {

  public static void show(String title, String message, Runnable runnable) {
    Stage dialogStage = new Stage();
    dialogStage.initStyle(StageStyle.UTILITY);
    dialogStage.setResizable(false);
    dialogStage.initModality(Modality.APPLICATION_MODAL);
    dialogStage.setTitle(title);

    BorderPane borderPane = new BorderPane();

    Label label = new Label(message);
    label.setWrapText(true);
    label.setPrefWidth(360);
    borderPane.setTop(label);

    ProgressIndicator progress = new ProgressIndicator(-1);
    borderPane.setCenter(progress);

    Scene scene = new Scene(borderPane);
    dialogStage.setScene(scene);

    dialogStage.show();

    Task<Void> task = new Task<>() {
      @Override
      protected Void call() throws Exception {
        runnable.run();
        return null;
      }
    };

    task.setOnSucceeded(event -> {
      dialogStage.close();
    });

    new Thread(task).start();
  }
}

File: src/main/java/ch/obermuhlner/astro/javafx/ProgressDialog.kt
package ch.obermuhlner.astro.javafx

import javafx.concurrent.Task
import javafx.concurrent.WorkerStateEvent
import javafx.event.EventHandler
import javafx.scene.Scene
import javafx.scene.control.Label
import javafx.scene.control.ProgressIndicator
import javafx.scene.layout.BorderPane
import javafx.stage.Modality
import javafx.stage.Stage
import javafx.stage.StageStyle

object ProgressDialog {
    fun show(title: String?, message: String?, runnable: Runnable) {
        runnable.run()
    }

    fun showREALLY(title: String?, message: String?, runnable: Runnable) {
        val dialogStage = Stage()
        dialogStage.initStyle(StageStyle.UTILITY)
        dialogStage.isResizable = false
        dialogStage.initModality(Modality.APPLICATION_MODAL)
        dialogStage.title = title
        val borderPane = BorderPane()
        val label = Label(message)
        label.isWrapText = true
        label.prefWidth = 360.0
        borderPane.top = label
        val progress = ProgressIndicator(-1.0)
        borderPane.center = progress
        val scene = Scene(borderPane)
        dialogStage.scene = scene
        dialogStage.show()
        val task: Task<Void?> = object : Task<Void?>() {
            @Throws(Exception::class)
            override fun call(): Void? {
                runnable.run()
                return null
            }
        }
        task.onSucceeded = EventHandler { event: WorkerStateEvent? -> dialogStage.close() }
        Thread(task).start()
    }
}